### DISCUSSION TITLE
Pyright 1.1.301 no longer recognizing modules in first-party library

### DISCUSSION THREAD

**Discussion by slobberchops** [microsoft/pyright#4870] [Q&A] [answered]
When upgrading to pyright 1.1.301 we have begun to encounter a problem whereby pyright will not recognize module that come from first-party libraries. For example, a source file (`src/trajector/api/apps/survey/rest/application.py`) in one of our repositories imports a module from a library that is downloaded and installed into the virtual environment. 

```python
from trajector.api.service import errors
```

While this does not trigger any errors in pyright 1.1.300, the following error message occurs:

```
 /<home>/project/trajector-cloud-api/src/trajector/api/apps/survey/rest/application.py:15:35 - error: "errors" is unknown import symbol (reportGeneralTypeIssues)
/<home>/project/trajector-cloud-api/src/trajector/api/apps/survey/rest/user_surveys.py
```

This seems like a bug, and it's not clear what information I should provide in a potential bug report. I'm uncertain what steps to take to make this reproducible.

Is anyone else experiencing this kind of problem with 1.1.301? Am I correct in assuming that this is related to importing a module from a first-party package?

**Comment by erictraut** [comment]
When you say "first-party library", are you talking about a library whose source code is included in your project? When you install this library, are you using an editable install? If so, make sure you read [this documentation](https://microsoft.github.io/pyright/#/import-resolution?id=editable-installs).

If you're not using an editable install, can you provide more details about what you mean by "first-party library" and how it is installed?

**Reply by slobberchops** [reply]
Apologies, I should have provided a better explanation.

My meaning of a "first party library" is a library that is produced inside my organization for my organization. The thing that I thought might be relevant is that both the project that is using the library and the library share the same package names. Both the project and the library put modules underneath the `trajector` package.  Sometimes they share the same sub-package such as `trajector.api`. In the example, the main project has a module `trajector.api.apps.survey.rest.application` while the library distribution has a module `trajector.api.errors`.

The library is packaged using poetry and downloaded from a private repository.

**Reply by erictraut** [reply]
Import resolution behaviors are very complex in Python, especially when there are overlapping namespaces. The behavior depends on a bunch of factors including the import search path ordering and whether the modules and submodules are traditional packages or implicit namespace packages (i.e. lacking an `__init__.py` file).

I don't think we made any changes in 1.1.131 that would affect import resolution, but it's possible that one of our changes had an unintended side effect.

To help diagnose the problem, I recommend adding `"verboseOutput: true"` to your pyrightconfig.json file (or `"verboseOutput = true"` to your pyproject.toml file). Then open the project and observe an import resolution failure. Look at the log output. You should see a bunch of detailed (verbose) output for each import resolution failure. The output will show which attempts pyright has used in each case to resolve the import. Let me know if that gives you any insights.

**Reply by slobberchops** [reply]
Thank you very much, I went ahead and ran that. The verbose output as specified does not appear to have any additional information about specific resolution failures. However, what I did was compared the verbose output from `1.1.300` to `1.1.301`. There is no significant difference except for two things:
* The order of some of the messages has changed. There are two source directories in our python project: `src` and `dev`. In `1.1.300` it searches `src` then `dev`, and in `1.1.301` it searches `dev` then `src`.
* The following message appears a few times in `1.1.300` but not in `1.1.301`:
```
    Looking in stubPath '/Users/rafe/project/trajector-cloud-api/typings'
    Attempting to resolve stub package using root path '/Users/rafe/project/trajector-cloud-api/typings'
    Attempting to resolve using root path '/Users/rafe/project/trajector-cloud-api/typings'
```

This makes me wonder if it's related to `https://github.com/microsoft/pyright/issues/4862`?

**Reply by erictraut** [reply]
You could confirm your theory by explicitly setting your `stubPath` to `"./typings"` rather than relying on the default. The bug you referenced affected only the command-line version of pyright, not the language server. Is that consistent with what you're seeing?

**Reply by slobberchops** [reply]
Apologies other stuff came up, and I am intending to revisit this.

**Reply by slobberchops** [reply]
Setting `stubPath` to `./typings` in my configuration file (`pyproject.toml`) does not fix the problem. It continues to occur in version 1.1.302 in addition to 1.1.301.

**Reply by erictraut** [reply]
Let's try this. Within the statement `from trajector.api.service import errors`, right click on `service` and choose `Go to declaration`. Where does that take you? I presume it's in a file called `trajector/api/service.py` or `trajector/api/service/__init__.py`. If so, does this file have a symbol called `errors`? If so, how is it declared? Is it in an unreachable conditional code path?

**Reply by slobberchops** [reply]
`errors` is a module in a file name `errors.py`.

**Reply by slobberchops** [reply]
I'm looking through changes between 1.1.300 and 1.1.301 which might be the cause of the condition I'm seeing. I am not familiar with pyright code and don't know typescript so I apologize for being naive here and there. The change that popped out at me when looking is https://github.com/microsoft/pyright/commit/f63069227c6b0845ab8420a8784e6b350484e456 .

Issue is: https://github.com/microsoft/pyright/issues/4827

A test I did seems to support that the condition is encountered specifically when aliasing an import symbol. I was able to modify the import in a way that appears to demonstrate that `pyright` is still able to understand the kind of library relationship that I have described.

For example the import below when used from a user repository of the library package caused `pyright` to fail:

```
from trajector.common import timing
```

If I convert it to the error condition clears up:

```
import trajector.common.timing
timing = trajector.common.timing
```

**Reply by erictraut** [reply]
If you believe there is a bug here and you're able to provide detailed repro steps, please open a bug report. You may need to create a temporary repo that provides the minimal, self-contained code.

**Reply by slobberchops** [reply, marked as answer]
In an attempt to be as helpful as possible, I went ahead and set up a repository with a library and a user of that library in order to demonstrate the issue before and after. In the course of that work I discovered that functions import from the client library were not being type checked at all in either pre-1.1.301 nor 1.1.301-on.

Long story short, through trial and error and reading more documentation I discovered that Pyright really doesn't like to deal with libraries that are not PEP-561 compatible. This is because our libraries do not contain any `py.typed` files. When adding `py.typed` to the library package not only does the aliased import start working but the until now unnoticed unchecked library types start working.

This leads me to conclude that this is not in fact a bug, and we had been relying on possibly undefined behavior before 1.1.301. I will work to correct our libraries to confirm that this is the case. I will note back here if this corrects the problem.

Thank you for your time. It was very helpful.

**Reply by slobberchops** [reply]
Yes, I can confirm that adding `py.typed` to the correct packages will resolve the issue I was experiencing.
